lib/output/switch.go

```diff
@@ -0,0 +1,341 @@
+	var err error
+	for i, oConf := range conf.Switch.Outputs {
+		outputs[i], err = New(oConf.Output, mgr, logger, stats)
+		if err != nil {
+			return nil, err
+		}
+		if oConf.Condition != nil {
+			oConf.cond, err = condition.New(*oConf.Condition, mgr, logger, stats)
+			if err != nil {
+				return nil, err
+			}
+			confs[i] = oConf
+		}
+	}
+
+	return newSwitch(outputs, confs, mgr, logger, stats)
+}
+
+func newSwitch(
+	outputs []types.Output,
+	confs []switchConfigOutputs,
+	mgr types.Manager,
+	logger log.Modular,
+	stats metrics.Type,
+) (*Switch, error) {
+	o := &Switch{
+		running:      1,
+		stats:        stats,
+		logger:       logger.NewModule(".broker.switch"),
+		transactions: nil,
+		outputs:      outputs,
+		outputNs:     []int{},
+		confs:        confs,
+		closedChan:   make(chan struct{}),
+		closeChan:    make(chan struct{}),
+	}
+
+	o.throt = throttle.New(throttle.OptCloseChan(o.closeChan))
+
+	o.outputTsChans = make([]chan types.Transaction, len(o.outputs))
+	o.outputResChans = make([]chan types.Response, len(o.outputs))
+	for i := range o.outputTsChans {
+		o.outputNs = append(o.outputNs, i)
+		o.outputTsChans[i] = make(chan types.Transaction)
+		o.outputResChans[i] = make(chan types.Response)
+		if err := o.outputs[i].Consume(o.outputTsChans[i]); err != nil {
+			return nil, err
+		}
+	}
+	return o, nil
+}
+
+//------------------------------------------------------------------------------
+
+// Consume assigns a new transactions channel for the broker to read.
+func (o *Switch) Consume(transactions <-chan types.Transaction) error {
+	if o.transactions != nil {
+		return types.ErrAlreadyStarted
+	}
+	o.transactions = transactions
+
+	go o.loop()
+	return nil
+}
+
+//------------------------------------------------------------------------------
+
+// loop is an internal loop that brokers incoming messages to many outputs.
+func (o *Switch) loop() {
+	defer func() {
+		for _, c := range o.outputTsChans {
+			close(c)
+		}
+		close(o.closedChan)
+	}()
+
+	var (
+		mMsgsRcvd  = o.stats.GetCounter("broker.fan_out.messages.received")
+		mPartsRcvd = o.stats.GetCounter("broker.fan_out.parts.received")
+		mOutputErr = o.stats.GetCounter("broker.fan_out.output.error")
+		mMsgsSnt   = o.stats.GetCounter("broker.fan_out.messages.sent")
+		mPartsSnt  = o.stats.GetCounter("broker.fan_out.parts.sent")
+	)
+
+	for atomic.LoadInt32(&o.running) == 1 {
+		var ts types.Transaction
+		var open bool
+
+		select {
+		case ts, open = <-o.transactions:
+			if !open {
+				return
+			}
+		case <-o.closeChan:
+			return
+		}
+		mMsgsRcvd.Incr(1)
+		mPartsRcvd.Incr(int64(ts.Payload.Len()))
+
+		var outputTargets []int
+		for i, oConf := range o.confs {
+			if oConf.cond == nil || oConf.cond.Check(ts.Payload) {
+				outputTargets = append(outputTargets, o.outputNs[i])
+				if !oConf.Fallthrough {
+					break
+				}
+			}
+		}
+		if len(outputTargets) == 0 {
+			select {
+			case ts.ResponseChan <- response.NewError(ErrSwitchNoConditionMet):
```

Comment:
I think in this event we should probably return `response.NewAck()` and be sure to expose the drop via metrics, it's possible that users might intentionally set up their outputs to entirely drop certain message types.

Reply:
I was hoping to find someway to avoid situations where messages are unintentionally dropped by forcing users to use some combination of `filter` processors and/or `/dev/null` outputs for messages where dropping is the desired behavior. what are your thoughts on adding an additional config field to tune this behavior, either `ack_unmatched: true`, `require_match: true`, `min_match: 1`, or something similar?

```yaml
output:
  type: switch
  switch:
    ack_unmatched: true
    outputs:
    - output:
        type: foo
        foo:
          foo_field1: blah
      condition:
        type: text
        text:
          operator: contains
          arg: foo
    - output:
        type: bar
        bar:
          bar_field1: blah
      condition:
        type: text
        text:
          operator: contains
          arg: bar
```

vs

```yaml
output:
  type: switch
  switch:
    outputs:
    - output:
        type: foo
        foo:
          foo_field1: blah
      condition:
        type: text
        text:
          operator: contains
          arg: foo
    - output:
        type: bar
        bar:
          bar_field1: blah
      condition:
        type: text
        text:
          operator: contains
          arg: bar
    - output:
        type: stdout
        processors:
          - type: filter
            filter:
              type: static
              static: true
```

Reply:
I'd prefer `ack_unmatched` (maybe instead called `drop_unmatched`) out of those options. But I'm also happy to leave it out entirely, I think it's reasonable to expect the messages to be filtered out if you haven't created an output that matches them.